lib/spack/spack/package.py

```diff
@@ -1799,7 +1805,7 @@ def build_process():
                 self, build_process, dirty=dirty, fake=fake)
 
             # If we installed then we should keep the prefix
-            keep_prefix = self.last_phase is None or keep_prefix
+            keep_prefix = self.last_phase is None or self.stop_before_phase is None or keep_prefix
```

Comment:
This needs to be

`keep_prefix = (self.last_phase is None and self.before_phase is None) or keep_prefix`